src/pybind/mgr/cephadm/services/ha_rgw.py

```diff
@@ -0,0 +1,161 @@
+import json
+import logging
+from typing import List, cast, Tuple, Dict, Any
+
+from ceph.deployment.service_spec import HA_RGWSpec
+
+from orchestrator import DaemonDescription, OrchestratorError
+from .cephadmservice import CephadmDaemonSpec, CephService
+from ..utils import CephadmNoImage, cephadmNoImage, resolve_ip
+
+logger = logging.getLogger(__name__)
+
+
+class HA_RGWService(CephService):
+    TYPE = 'HA_RGW'
+
+
+class HAproxyService(CephService):
+    TYPE = 'haproxy'
+
+    class rgw_server():
+        def __init__(self, hostname: str, address: str):
+            self.name = hostname
+            self.ip = address
+
+    def prepare_create(self, daemon_spec: CephadmDaemonSpec[HA_RGWSpec]) -> CephadmDaemonSpec:
+        assert self.TYPE == daemon_spec.daemon_type
+        assert daemon_spec.spec
+
+        daemon_id = daemon_spec.daemon_id
+        host = daemon_spec.host
+        spec = daemon_spec.spec
+
+        out, err, code = self.mgr._run_cephadm(host, cephadmNoImage, 'verify-prereqs',
+                                               ['--service-type', 'haproxy'],
+                                               no_fsid=True, error_ok=True)
+
+        if code:
+            raise OrchestratorError(
+                'HAproxy failed prereqs check with code: %d, stderr:%s' % (
+                    code, '\n'.join(err)))
```

Comment:
Exactly. at that point you can't do anything about it, except for raising something.